### ISSUE TITLE
Add telemetry info to image metadata

### ISSUE THREAD

**Issue by nathanchan631** [uas-at-ucla/uas-2024#21]
When we send a request to the vision server, we want to bundle the telemetry data and the image together (by adding the telemetry data to the image metadata) so we do not need to make two separate requests. How can we do that?

**Comment by kushhansagarwal** [comment]
From `suas-onboard-2022`, ODLC is posted as following

```
def queue_image_for_odlc():
    """
    Queue image POST request
    """
    # Save file locally, so we can process it using OpenCV
    raw_data = request.get_data()
    file_location = f'{FILE_PATH}/{time.time_ns()}-'
    with open(file_location, 'wb') as file:
        file.write(raw_data)
        image_queue.put({"file_location": file_location,
                         "telemetry": drone.get_telemetry()})

    return Response(status=200)
```
where `drone.get_telemetry()` gets the JSON telemetry data. The `pyexif` library can be used to get the EXIF data and edit it as follows

```
image_path = "./data/IMG_0073.JPG"
metadata = pyexif.ExifEditor(image_path)
metadata.setTag('OrientationX', '-12.1')
```

The telemetry data, presumably in JSON format be iterated over and appended as separate EXIF tags, or possibly a single JSON string which can then be parsed when needed.

```
import pyexif as exif
file_location = f'{FILE_PATH}/{time.time_ns()}-'
image_exif = exif.ExifEditor(file_location)

telemetryData = drone.get_telemetry()

for dataField in telemetryData:
    image_exif.setTag(dataField, dataField.value)

```

The exact **format of the telemetry data has to be decided** and known before saving it on EXIF.

[Reference](https://www.linkedin.com/pulse/manipulating-image-exif-data-python-natasha-kacoroski/)

**Comment by nathanchan631** [comment]
@kushhansagarwal 

Good work, this is great progress. Just a quick note:

This is the new `queue_image_for_odlc` function: https://github.com/uas-at-ucla/uas-2024/blob/main/vision/main.py#L67

I'm thinking the camera script (flight software, not vision) will save the images, along with the telemetry data in the metadata. Then, we just pass the image path to vision, and vision reads the image path from disk for odlc.

Last year, we were making two separate HTTP requests, one to post telemetry, and one to post the image. This year, we are just going to make just one post request for the image, which will have the telemetry included in the exif. What do you think?

**Comment by kushhansagarwal** [comment]
I think it sounds great, I can write some test stuff and see what performance improvements we are getting with this. I am not sure how efficient or fast `pyexif` is, so will need to figure that out.

Also, what kind of telemetry data would we be needing? Since we are making a drone, just the exact GPS of the spot where the photo was taken should suffice right? Or are we doing a loop which goes something like this

1. Flight software takes photo at the interval
2. Vision identifies a target in let's say top right of the image 
3. Vision knows the GPS coordinates of position, POV of camera, position of target in image, and can calculate the GPS coordinates of the target (different from that of the photo since it is not centered)
4. Flight software navigates to the new location and clicks another photo
5. Vision verifies that the target is not indeed in the center of the frame
6. Flight software triggers the drop mechanism

Since we have the 25ft radius requirement for a drop to count successful, I think this may be ideal. In that case, I imagine the flow to go as follows

1. Flight software snaps photo, POST to queue
2. Vision scans through photos, marks photos which targets that match the ones we have to drop on
3. Computes the exact GPS of the target as mentioned above
4. Flight software receives the location for the drop
5. It navigates to the location and drops it

I don't know if we are still using parachute (which may need wind data too) or something like a winch where we lower the object and drop it there. If we are doing the winch thing, we can have a camera on the payload assembly thats lowered and skip computing the exact GPS coordinate or we can have an AruCo on the payload and use the main camera to correct our drop location.

**Comment by kushhansagarwal** [comment]
**UPDATE: @nathanchan631  I have failed**

EXIF is being funny and I am not able to store data to it easily. Instead since we already have redis running and the images are being put in the queue already, we can just add another field with the metadata in JSON format. This should improve performance too.

When targets are detected in an image, the call to the localization function can send the redis uid of the image as well and the function can pull the file location and the metadata at that time.

**Comment by nathanchan631** [comment]
What exactly are you running into with exif? Using redis is a great backup, but not sure I want to default to that just yet.

**Comment by kushhansagarwal** [comment]
> My thought atm is I just want to avoid polluting redis with data that is only used once and then deleted.

I am not sure if I understand how it worked last time correctly, but I am assuming that there is an image queue which stores the file paths in Redis. If an image has detections which we have to drop on, we send it over for the target to be localized. Since the file path is already going into the Redis queue, we can add the metadata to the same object and retrieve it in the localization part of the code. Can you tell me if this is how it works? I will review last time's code too to get an understanding as well

---

The main issue that I am running into is that EXIF can be altered in two ways

- Using the `exiftool` CLI
- Reading an image using Python, altering the metadata, save it

`exiftool` can be called in Python using the `subprocess` library, I have not run any tests yet but I don't think it mould be more efficient than using an EXIF library in Python. The issue I am getting with EXIF library is that I am able to open the image, add custom metadata to it and I am able to read it before saving.

Once I save it, any custom tags which are added are lost. I can save custom information in existing EXIF tags though, so if we want to proceed that way we can do that.

In essence,

```python
with open(image_path, 'rb') as image:
    my_image = Image(image)

    payload_json = {
        "altitude": 100,
        "latitude": 34.06901102425351,
        "longitude": -118.4452,
    }
    my_image.set('Payload', str(payload_json))

    # If I read the tag here, it works perfectly

    with open('modified.jpg', 'wb') as new_image:
        new_image.write(my_image.get_file())

        # If i try to read the EXIF on new_image, I am not able to access the custom tag, 
        # only the existing ones. However, if I alter an existing tag such as Model, I can access the updated data
```
I am not sure about the performance differences, but I can surely index them and let you know. I would be comparing these two processes and see the total time. Let me know if any other metrics are to be checked too. I am not including detections here as I assume they would be handled by a different part of the code which would be the same irresepctive of how we handle metadata

**Saving using EXIF**

- Get the telemetry data which is to be appended
- Read an image on disk with no custom metadata
- Append custom metadata to it on the `ImageDescription` EXIF field
- Save the image
- Open the modified image and read the metadata

**Saving using Redis**

- Get the telemetry data which is to be appended
- Save the data in the same document/[object](https://redis.io/docs/data-types/json/) which matches the file path
- Read the data

I don't have a lot of experience with Redis, but I will get an instance running and profile the performance

**Comment by nathanchan631** [comment]
The more I am thinking about it, the more I'm realizing Chirag and I way overcomplicated the system last year lmao. I'll spare you the details.

New plan:
1. Flight software script takes a photo, saves it in a subfolder within the vision directory
2. Flight software script also gets telemetry data and makes a single HTTP request to the vision endpoint, let's call it POST odlc. It will be a json blob with three fields: `image_path`, `latitude`, `longitude`.
3. Vision receives the request and adds it to a queue. I've already started implementing this here: https://github.com/uas-at-ucla/uas-2024/blob/main/vision/main.py
4. A separate vision script will pop from the queue and send it through the pipeline
5. Any possible detections will be added to redis.
6. Once the queue is empty, vision will set its status to done.
7. This whole time, ever since the airdrop scan is done, flight software has been sending a GET status request to vision. Once vision returns that the queue is done, flight software will send a GET odlc request.
8. Vision will then calculate the top N most probable detections and return it to flight software
9. Flight software does airdrop

This is probably more information than you needed to know, but just wanted to give it all to you so you so you could point out anything you think could be improved.

**Comment by kushhansagarwal** [comment]
This looks great. Have we finalised this approach? If so, I can get started on it.

If we haven't can we try messing around with WebSockets? I think it would be a cleaner implementation and would eliminate the need for Redis entirely. I can explain my implementation and let me know if it is viable

## Potential WebSockets implementation

The whole mission can be divided into four segments (essentially microservices)

- Flight software: Flying the drone
- Camera: Taking photos and telemetry
- Detector: Running ODLC on images
- Drop: Dropping the payload on trigger

All these segments would be hooked to the same WebSockets server running locally. As the drone flies over the targets area, flight software would be taking photos. The camera microservice would post such objects on the WS channel

```JSON
{
    "clientId" : "bbaf7954-6fbc-11ee-b962-0242ac120002", 
    "filePath": "images/000001243.jpg",
    "location": {
        "latitude": 37.7749295,
        "longitude": -122.4194155
    },
    "telemetry": {
        "orientation": {
            "x": 0.000000,
            "y": 0.000000,
            "z": 0.000000,
        },
        "acceleration": {
            "x": 0.000000,
            "y": 0.000000,
            "z": 0.000000,
        }
    }
}
```
The detector microservice can have a buffer which would store these messages, eliminating the need for Redis and another script to push it to the pipeline. Upon receiving the message, it would start ODLC on it. Once it has completed, it can post an object like

```json
{
  "clientId" : "f047b508-6fbe-11ee-b962-0242ac120002", 
  "detections": [
    {
      "type": "alphanumeric",
      "shape": "star",
      "color": "red",
      "letter": "A",
      "confidence" : 0.99,
      "location": {
        "latitude": 37.7749295,
        "longitude": -122.4194155
      }
    }
  ]
}
```

This message would be received by the drop microservice which would already have the list of targets. It would keep a record of the most confident detection for each of the targets. When flight software decides that its time to drop, it can just send a message on the WS channel instructing to begin the drop. To do so, the drop microservice would send the target location to flight software (over WS), flight software would go their, signal once it is there, and then the drop microservice can execute the drop.

---

I think this would work great because

- Data which is not needed would not be stored (as you had mentioned)
- Debugging becomes easier since all the communications being sent can be viewed and debugged
- The GCS display requirement in SUAS specifics can be easily satisfied, since all the communication can be tapped by adding another client to the WS channel. We can also send messages from the GCS (such as the order in which the water bottles will be kept since each water bottle has the target details printed on it)

I have worked with WebSockets before and in my experience they have worked great every single time. For IDEA Hacks 2023, my team and I had implemented WebSockets for our project Cryptopoly where 5 services were hooked on the same channel including a script running CV, a dashboard and a matrix display. It worked great, was super fast and three of the services were performing insanely well and quickly despite running on an ESP32.

I have also implemented it [here (desktop only)](https://e96planes.vercel.app) (the WS channel is down right now, but you can simulate it by hitting the yellow generate button. The left would display the raw data and you can see the parsed information on the right). This is a visualizer for E96 Drones and E96 Planes where a GPS + Pico W payload on the aircraft constantly broadcasts the information to the WS channel (on a server on the internet) and the website parses and displays it. This can be used as our GCS too since we have all the data anyways.

**Comment by nathanchan631** [comment]
I'm not too familiar with websockets, but you seem excited to try it. I feel like it may be overkill, but if you want to do it, I guess there's no reason not to, so go ahead. You can create your own task for it if you want, but doesn't really matter since you'll likely be the only one working on it lol.

The only thing is that I don't think you need separate microservices for the camera and airdrop, since both are just single functions we call from the main driver script.

> The GCS display requirement in SUAS specifics can be easily satisfied, since all the communication can be tapped by adding another client to the WS channel. We can also send messages from the GCS (such as the order in which the water bottles will be kept since each water bottle has the target details printed on it)

I haven't set it up yet (so maybe it's a pain in the ass) but the drone software we use has an option to add a GCS. If you want to use the websocket for ground control, we would have to make our own UI right?

**Comment by kushhansagarwal** [comment]
Yayyy I will get to it. WS is basically a pipeline between clients where anyone can send data to anyone. Each payload is tagged with a recipient uid so each client knows which data is intended for them (since everyone receives all messages).

Since WS is kind of an alternative to HTTPS, for now I would call the HTTPS endpoints under the hood when a task is to be done, but later on we can remove the wrapper and possibly see performance improvements.

>  I don't think you need separate microservices for the camera and airdrop

The only reason I would be do that is because they act as separate codebases. Maybe its incorrect to term them as microservices but the camera code and airdrop code would connect to the WS channel separately, since the camera would only be posting data to the channel while airdrop would be receiving it. Alternatively, we can have flight software directly trigger the drop.

I have not thought about the organization of it completely, I can make a flowchart of sorts soon so it is clearer.

> we would have to make our own UI right?

Yes, but I won't worry about it too much though. I have made a lot of stuff in Svelte + DaisyUI and they give good looking and fast websites. The [website I linked earlier](https://e96planes.vercel.app/) is also made completely with SvelteKit and didn't take too long.

> drone software we use has an option to add a GCS

From what I remember it only shows you the map and telemetry right? If we make our own we can also look at all the messages in real time and visualize everything. Depending on how we communicate, we can also relay the images, detections etc. IIRC last year we only had numerical data and not detections and stuff in the front end.

**Comment by nathanchan631** [comment]
If the UI isn't too much work, then sure. Also we had a way of accessing detections through redis last year, just don't know if it was ever implemented in the front end.

> The only reason I would be do that is because they act as separate codebases.

I'm not completely sure what you mean by codebase, but you can look at the code from last year. Airdrop is just one function that is passed detection locations, and camera is just one function that send requests to vision server.

I'm not sure about the exact implementation, but I feel like a system where the camera function sends image through websocket and flight software driver code gets detections and then can tell airdrop script where to go, is easier.

**Comment by kushhansagarwal** [comment]
> I'm not sure about the exact implementation, but I feel like a system where the camera function sends image through websocket and flight software driver code gets detections and then can tell airdrop script where to go, is easier.

Hmm I was thinking of a slightly different approach I think. I will do one thing, I will send a flowchart detailing all the processes and data exchanges. Even if we want to change something in the future, it would be pretty easy with the WS backbone set up.

**Comment by kushhansagarwal** [comment]
@nathanchan631  I have added a [description](https://github.com/uas-at-ucla/uas-2024/blob/kush/vision/websockets/README.md) here with how WebSockets would implement the mission. Let me know if you have any questions and if it is viable to implement.

**Comment by nathanchan631** [comment]
Ah, now I see. The reason you want to have more than just two websocket connections is to make code more modular. If that causes no issues with traffic or processing load, then sounds good. I have a few small questions, but I guess it's best to wait to see your implementation.

However, your implementation deviates from last year's in a significant way. In your way, we deliver and scan airdrops concurrently, but in last years', we scanned the entire area first, and then used a matching algorithm to get our final detections. In your way, how do we guarantee that we scan the entire area without missing any targets? However, which is better ultimately depends on the accuracy of our vision system, so might be too early to make the call right now. This decision can always wait anyway.

I'm excited to see how this turns out! Once you're done implementing it, make sure to add it to add a command to start the websocket to the makefile.

**Comment by kushhansagarwal** [comment]
Perfect, I'll get started on it. 

> In your way, we deliver and scan airdrops concurrently

Oh I don't think it was very clear in the doc, I'll update it. But in this we still scan first and drop later. Since we are given a boundary for the zone where targets would be, flight software would first fly over the zone while simultaneously clicking photos. Once the zone is covered, we would have the most confident list of targets and then the drone would navigate to them and drop. 

However, we **can** implement the scan and drop concurrently if we want to do it quicker. We can put in a confidence threshold of let's say 0.95 and if a target is detected above that confidence it would immediately drop. 

I'd make it so that such changes are easy to implement at a later date so we can see what works out best for us. 

For now, I'd keep everything in a folder called `websockets` and make different files for the microservices which would be wrappers on the code in the flight and vision folders. Therefore, others can work in the existing directories and if we end up not using WebSockets, no extra work has to be done

**Comment by nathanchan631** [comment]
Sounds good.
